### PR TITLE
refactor(decode): return errors instead of panic

### PR DESCRIPTION
**Pull request by nicksinch**
**Detailed description**:
- refactors all **decode** methods on all types to return an error if decoding goes wrong instead of panic 

**Which issue(s) this PR fixes**:
part of #56 

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: boolean_test.go [hunk 1]
```diff
@@ -44,7 +45,7 @@ func Test_DecodeBool(t *testing.T) {
 			buffer := &bytes.Buffer{}
 			buffer.Write(testExample.input)
 
-			result := DecodeBool(buffer)
+			result, _ := DecodeBool(buffer)
```
Comment (radkomih) [inline boolean_test.go:48]: assert that there is no error

File: common.go [hunk 2]
```diff
@@ -46,7 +51,11 @@ func decodeByType(i interface{}, buffer *bytes.Buffer) Encodable {
 	case Compact:
 		return DecodeCompact(buffer)
 	case Sequence[U8]:
-		return Sequence[U8](DecodeSliceU8(buffer))
+		dec, err := DecodeSliceU8(buffer)
+		if err != nil {
+			return Empty{}, err
```
Comment (radkomih) [inline common.go:56]: here you can return `nil, err`

File: common.go [hunk 3]
```diff
@@ -55,7 +64,7 @@ func decodeByType(i interface{}, buffer *bytes.Buffer) Encodable {
 	// case Result[Encodable]:
 	// return DecodeResult(buffer)
 	default:
-		panic("type not found")
+		return Empty{}, errTypeNotFound
```
Comment (radkomih) [inline common.go:67]: `nil, errTypeNotFound`

File: compact_test.go [hunk 4]
```diff
@@ -82,7 +82,7 @@ func Test_Decode_Compact(t *testing.T) {
 			buffer := &bytes.Buffer{}
 			buffer.Write(e.input)
 
-			result := DecodeCompact(buffer)
+			result, _ := DecodeCompact(buffer)
```
Comment (radkomih) [inline compact_test.go:85]: assert that there is no error

File: dictionary_test.go [hunk 5]
```diff
@@ -55,7 +55,7 @@ func Test_DecodeDictionaryStrBool(t *testing.T) {
 			buffer := &bytes.Buffer{}
 			buffer.Write(testExample.input)
 
-			result := DecodeDictionary[Str, Bool](buffer)
+			result, _ := DecodeDictionary[Str, Bool](buffer)
```
Comment (radkomih) [inline dictionary_test.go:58]: assert that there is no error

File: dictionary_test.go [hunk 6]
```diff
@@ -115,7 +115,7 @@ func Test_DecodeDictionaryU8Str(t *testing.T) {
 			buffer := &bytes.Buffer{}
 			buffer.Write(testExample.input)
 
-			result := DecodeDictionary[U8, Str](buffer)
+			result, _ := DecodeDictionary[U8, Str](buffer)
```
Comment (radkomih) [inline dictionary_test.go:118]: assert that there is no error

File: i128_test.go [hunk 7]
```diff
@@ -67,7 +67,7 @@ func Test_DecodeI128(t *testing.T) {
 			buffer := &bytes.Buffer{}
 			buffer.Write(e.input)
 
-			result := DecodeI128(buffer)
+			result, _ := DecodeI128(buffer)
```
Comment (radkomih) [inline i128_test.go:70]: assert that there is no error

File: i8.go [hunk 8]
```diff
@@ -12,7 +12,11 @@ func (value I8) Bytes() []byte {
 	return U8(value).Bytes()
 }
 
-func DecodeI8(buffer *bytes.Buffer) I8 {
+func DecodeI8(buffer *bytes.Buffer) (I8, error) {
 	decoder := Decoder{Reader: buffer}
-	return I8(decoder.DecodeByte())
+	db, err := decoder.DecodeByte()
```
Comment (radkomih) [inline i8.go:17]: We should try to keep things consistent, in one place the return number is `dec64/dec32/...`, here is `db`.
Why not `dec8`?

File: option_test.go [hunk 9]
```diff
@@ -889,9 +889,8 @@ func Test_DecodeOptionPanicsMissingBoolBytes(t *testing.T) {
 			buffer := &bytes.Buffer{}
 			buffer.Write(e.input)
 
-			assert.Panics(t, func() {
-				DecodeOption[Bool](buffer)
-			})
+			_, err := DecodeOption[Bool](buffer)
+			assert.ErrorContains(t, err, errors.New("EOF").Error())
```
Comment (failfmi) [inline option_test.go:893]: `errors.New("EOF")` -> `io.EOF`

**Review by failfmi — APPROVED** [review]

**Review by radkomih — APPROVED** [review]

**Pull request merged**